## RSA - The Project

Name:  Angela Booth

Moodle Email: anbo3523@colorado.edu

Name your file "cuidentikey".ipynb  This means YOUR identikey. 

**Do NOT delete instructions or pre-existing comments in the code. Removal will trigger significat point reductions** 

**Required: Add your own comments as if the instructions and comments are not visible - this allows you to customize your project after the course**



**The Prompt:**

Your boss has been long intrigued by the RSA algorthim and has asked you to create a turorial RSA walk through and exploration project to help the boss and others in the company to understand the basic structures and mathematics of an RSA implementation.

Your project will:
1. Implement the algorithm for breaking codes with a given basic structure. 
2. Explain the steps in an "Inverview Grading Style" with Jupyter Lab Text blocks demostrating and explaining the function as you go. 
3. These demonstrations, explanations, and code comments must demonstrate that you understand the code and what it is doing. 
4. After the basic implementation you will rewrite the code completely with your own improvements, or copy the basic code package and add new features. This will be your custom feature.
5. IMPORTANT - in the first implementation you will keep all instructions, given comments in the code, and use the variable names and set up as given. For your custom feature you may remove these and make the code "your own." ALL CHANGE must be explained in text - why did you change/make improvements? We should be able to see how you adapted the orginal code.






<hr />

# Table of Contents/Grading Structure

#### 0. Bonus +/- Overall structure, code comments, and following directions. (Score may go up or down if overall structure or format is excellent or incomplete). Not extra credit.

## 1. Introduction to your RSA Project (5 points)


## 2. Code Package: FME    (10 points)
        * Convert_Binary_String 
        * FME

## 3. Code Package: Key Generation using Euclidean Algorithms    (10 points)
        * Euclidean Algorithm
        * Extended EA
        * Find_Public_Key_e
        * Find_Private_Key_d

## 4. Code Package: En/Decode Your Messages  (10 points)
        * Convert_Text
        * Convert_Num
        * Encode
        * Decode


## 5. Create a Code Demo(10 points)

## 6. Code Exchange Results  (5 points)

## 7. Write a Narrative (25 points)

## 8. Code Breaking:  (10 points)

## 9. Code Breaking: Complete Examples  (5 points)

## 10. Custom code feature / Advanced options  (10 points) 


    


## 1. Introduction to your RSA Project (5 points) 
<span style="color:DarkRed">Write an introduction/overview here. This should be something someone who has never seen your project could read to understand what is coming. Give a broad summarized overview of RSA components, discuss your custom feature a bit, and  include some preliminary narrative regarding your process. Much of this will be fleshed out in your narrative below, so remember to simply summarize details. This should be a robust paragraph or two of 200 - 300 words.</span>


In this project, you will be guided through the entire process of RSA (Rivest, Shamir, and Adleman) encryption and decryption. In the beginning of this project, we will be outlining all the functions required to make this algorithm work. The functions we will be going over include the following: Convert_Binary_String, FME, Euclidean_Alg, EEA, Find_Public_Key_e, Find_Private_Key_d, Convert_Text, Convert_Num, Encode, and Decode. In each of the functions, there will be notes that go over what their purpose is and how they work to get the values we need. We will then generate our own public and private keys using the functions we created. Additionally, we will demonstrate the use of these functions by encoding our own message with our own prime numbers by going through each function we use, why, and the outputs that we receive. Then, we continue with the decoding process and demonstrate how we use the private key to decode our messages. We will also show additional examples of encoding and decoding by using examples provided by our peers. Finally, you will see how we can break the code using brute force and just the public key. You will find that smaller numbers will be quite easy to break, proving the importance of having a long key.  

## 2. Code Package FME (10 points)

<span style="color:DarkRed"> 

FME is an essential function for RSA to pre-process the messages before the messages are encoded and decoded by the RSA algorithm. 
You have 3 choices of algorithms for this:
* The Rosen Book FME - which will need Convert_Binary_String below.
* Sriram's FME from the video.
* Your own "Square and Mod" function defined with a look up table.
    
*Do not copy code from the internet - it is recognized immediately by our software. In particular, use of bitshifting is forbidden in the first part of the project but may be explored in the custom feature.*
</span>



In [71]:
def Convert_Binary_String(_int):
    """
    Here, you need to define a function that converts an integer to
    a string of its binary expansion. You may or may not need this function. 
    
    For example:
    _int = 345
    bits = 101011001
    """
    #This function converts integers into binary expansion by dividing the integer by 2 and then taking the remainder and adding that to a binary string.
    #The remainder on each iteration represents the binary digits.
    #Here, we return 0 if the integer is 0. This is because the integer 0 in binary is equal to 0.
    if _int == 0:
        return "0"
    binary_str = ""
    #We are creating a loop here that will run until the integer becomes 0.
    while _int > 0:
        remainder = _int % 2 #Here, we are obtaining the remainder of the integer when moded with 2. This will give us a 0 or 1.
        binary_str = str(remainder) + binary_str #Take the remainder we received above and add it to the beginning of binary_str.
        _int = _int //2 # We need to update the integer to then continue finding the next binary digit.
    return binary_str #Once _int = 0, we have our complete binary expansion for the integer
#return bin(_int)[2:] this would be a shorthand way to take int to binary

In [3]:
def FME(b, n, m):
    """
    1. Using the fast modular exponentiation algorithm,
    the below function should return b**n mod m.
    As described on page 254. (however, you may input the exponent and 
    then convert it to a string - the book version imports the binary expansion)  
    2. You should use the function defined above Convert_Binary_String() if using the book algorithm.
    3. For this block you MUST use one of the 3 methods above.
    4. Any method using bit-shifting or copied from the internet (even changing varibale names) will result in a 0.
    
    **If you are completely stuck, you may use pow() with a 10pt penalty.**

    You may use the function you developed in your Mastery Workbook - but be sure it is your own
    work, commented, etc. and change inputs as needed.
    """  
    #This is the Fast Modular Exponentiation Algorithm
    #I am creating an algorithm that uses the method outlined in the book to calculate the mod of a number that includes a base and exponent (b^e)
    #In Discrete Mathematics and Its Applications book: ALGORITHM 5 Modular Exponentiation
    #b = base, n = exponent, and m = modulus
    #Through our Mastery workbook, we learned that FME is a lot more efficient than non-FME
    #Below, you will see the FME vs non-FME times that I calculated:
    #Varying n, fixed a:
    #n = 1000: NOT_FME time =  0.000004s, FME time = 0.000003s
    #n = 10000: NOT_FME time =  0.000032s, FME time = 0.000002s
    #n = 100000: NOT_FME time =  0.000447s, FME time = 0.000003s
    #n = 1000000: NOT_FME time =  0.006290s, FME time = 0.000004s

    #Varying a, fixed n:
    #a = 10: NOT_FME time = 0.000017s, FME time = 0.000002s
    #a = 100: NOT_FME time = 0.000014s, FME time = 0.000002s
    #a = 1000: NOT_FME time = 0.000053s, FME time = 0.000002s
    #a = 10000: NOT_FME time = 0.000043s, FME time = 0.000002s
    result = 1
    
    while n > 0:
        #If bit of exponent is odd, multiply base with result. We do this to isolate the power of the base corresponding to the bits that are 1
        #ex: 5^13 mod 17. 13 in binary is 1101. (Starting from the right) 1 = include 5^1, 0 = exlude 5^2, 1 = include 5^4, 1 = include 5^8.
        if (n % 2) == 1: #Checking if the mod of the number is 1, if so, it must be 1. 0 would return a mod of 0
            result = (result * b) % m
        
        #Square the base
        b = (b * b) % m
        
        #Divide the exponent by 2
        n = n // 2
    #Once e = 0, we will return the value, which is the answer of the b^e % m     
    return result

## 3. Code Package: Key Generation using Euclidean Algorithms    (10 points)

<span style="color:DarkRed">

The functions below will generate the public and private key pairs which will then be used to create a ciphertext using the public key and then decode the same using the private key. </span>



In [28]:
def Euclidean_Alg(a, b):
    """
    1. Calculate the Greatest Common Divisor of a and b.
    
    2. This version should have only positive inputs and outputs.
    
    3. The function must return a single integer ('x') which is
    the gcd of a and b.
    
    
    """
    #Here we are implementing the Euclidean Algorithm.
    #In order to find the GCD, we will want to continously take the mod of the two integers (a and b)
    #until we reach a mod of 0
    #This loop works by taking the mod of a and b, storing it in b, and taking b and storing that in a.
    #Then when the loop runs again, the process continues until we reach b (the mod) = 0
    #We can then return a which will be the GCD
    #We will be using this later to find whether two values are coprime (gcd of a and b = 1)
    #inspired by pseudocode on pg 269 of the "Discrete Mathemetics and It's Applications" book by Kenneth Rosen
    while b != 0:
        a, b = b, a % b 
    return a

In [29]:
def EEA(a, b):
    """
    This is a helper function utilizing Bezout's theorem as discussed in your MW.
    You will follow these same steps closely to construct this function.
    
    This version will return both: 
    1. the GCD of a, b 
    2. Bezout's coefficients in any form you wish. We recommend returning your coefficients as a list or a tuple. 
    HINT: return GCD, (s1, t1)
    
    * Ensure that your inputs are positive integers. Implement these kinds of checks.
    * It might also behoove you to consider reassigning a, b to new coefficients depending on which is greater.
    
    """
    #This function uses Extended Euclidean Algorithm to compute the GCD of two integers, a and b.
    #It also finds the Bezout coefficient s and t.
    # After we comeplete the algorithm, the value for s1 becomes our 'd' in our private key.
    
    #Initial values representing (s1, t1) and (s2, t2)
    s1, t1 = 1, 0 #represents coefficient for 'a'
    s2, t2 = 0, 1 #represents coefficient for 'b'
    
    #initial values of m and n
    m, n = a, b
    
    #While loop that will run until n becomes 0
    while n != 0:
        #k is the remainder of the division of m and n
        k = m % n
        #q is the quotient of the division of m and n
        q = m // n
        
        #update m and n with values we got above
        m, n = n, k
        
        #update s1, t1, s2, t2 (coefficients)
        new_s1, new_t1 = s2, t2
        new_s2, new_t2 = s1 - q * s2, t1 - q * t2
        
        #Assign new values to s1, t1, s2, t2
        s1, t1, s2, t2 = new_s1, new_t1, new_s2, new_t2
        
    #The final values of s1 and t1 are the Bezout coefficients
    #The values of m after the loop is the gcd of a and b
    return m, s1, t1
        
    
    

In [25]:
def Find_Public_Key_e(p, q):
    """
    Implement this function such that
    it takes 2 primes p and q.
    
    Use the gcd function that you have 
    defined before.
    
    The function should return 2 elements as follows:
    public key: n
    public key: e
    
    HINT: this function will run a loop to find e such 
    that e is relatively prime to (p - 1) (q - 1) 
    and not equal to p or q.
    
    NOTE: There are a number of ways to implement this key feature. 
    You, as the coder, can choose to how to acheive this goal.
    
    """
    #This function calculates the public key values which include n and e.
    #n is calculated by taking the product of the two prime numbers
    #e is calculated by first calculating phi (Euler's function) and then calling our Euclidean_Alg function which will provide us with our e
    
    n= p * q #Product of two prime numbers
    phi = (p - 1) * (q - 1) #Euler's function
    
    #Looping to find e such that 1 < e < phi and gcd(e, phi) = 1
    #Just as we did in our MW, we start at 2 and compare the value to phi, once we get a GCD of 1, we return those two values.
    for e in range(2, phi):
        if Euclidean_Alg(e, phi) == 1: #This finds whether our e and phi values are coprime.
            return n, e

In [26]:
def Find_Private_Key_d(e, p, q):
    """
    Implement this function to find the decryption exponent d, 
    such that d is the modular inverse of e. 
    
    This will use the Extended Euclidean Algorithm
    
    This function should return the following:
    d: the decryption component.
    
    This is not a single action, and there are multiple methods to create this. 
    
    You may create a helper function or have all code within this function.
    
    Plan ahead before coding this.

    """
    #This function calculates the private key exponent d for RSA decryption.
    #First, we calculate PHI (Euler's function) finds the number of integers that are coprime with p * q. 
    #Then we use the Extended Euclidean Algorithm to find d, the modular inverse of e mod phi
    #This will ensure e * d = 1mod phi.
    
    phi = (p - 1) * (q - 1)
    gcd, s, t = EEA(e, phi) #Extended Euclidean Algorithm to obtain d
    #the value for 'd' MUST be positive. Here we are checking for that
    #If not positive, we add 1 to 'd'
    if s < 0:
        s += phi
    return s

## 4. Code Package: En/Decode Your Messages  (10 points)

<span style="color:DarkRed">

1. In this part, you will define two functions `Encode` and `Decode` which will use the public and private keys that you calculated above.
2. Using the public key, the `Encode` function will encode a message and generate the corresponding cipher_text.
3. Using the private key, the `Decode` function will decode a cipher_text and recover the original message.
4. We are proving Convert_Text and Convert_Num for you.</span>



In [9]:
def Convert_Text(_string):
    """
    Define this function such that it takes in a simple 
    string such as "hello" and outputs the corresponding
    standard list of integers (ascii) for each letter in the word hello.
    For example:
    _string = hello
    integer_list = [104, 101, 108, 108, 111]
    
    You may use "ord()"
    """
    integer_list = [ord(char) for char in _string]#use ord to get ASCII value for each character
    return integer_list

In [6]:
def Convert_Num(_list):
    """
    Do the opposite of what you did in the Convert_Text
    function defined above.
    
    Define this function such that it takes in a list of integers
    and outputs the corresponding string (ascii).
    
    For example:
    _list = [104, 101, 108, 108, 111]
    _string = hello
    """
    _string = ''
    for i in _list:
        _string += chr(i)
    return _string

In [11]:
def Encode(n, e, message):
    """
    Here, the message will be a string of characters.
    Use the function Convert_Text from 
    the basic tool set and get a list of numbers.
    
    Encode each of these numbers using n and e and
    return the encoded cipher_text.
    """
    #This function will take a tring of characters, and convert them into ASCII.
    #Then we take those values and encode them by doing the following: (value we found in ASCII) ^ e % n. 
    #We will loop through this through the whole integer list
    integer_list = Convert_Text(message) #Convert message into ASCII
    #Encode each ASCII value using our public key
    cipher_text = [FME(num, e, n) for num in integer_list] #We are using FME (instead of pow). Check notes in FME function for functionality
    return cipher_text

In [1]:
def Decode(n, d, cipher_text):
    """
    Here, the cipher_text will be a list of integers.
    First, you will decrypt each of those integers using 
    n and d.
    
    Later, you will need to use the function Convert_Num from the 
    basic toolset to recover the original message as a string. 
    
    """
    #Here, we are taking the cipher_text and using our private key to decrept each of the integers.
    #In order to do this, we go through each value in cipher_text and then take that number and take it to the power of "d" and then mod it by n.
    #This will give us the ASCII value of our message, we then need to take that and covert it back to our message.
    decrypted_numbers = [FME(num, d, n) for num in cipher_text] #We are going through each value in the encoded text and using our private key to decrept it
    #convert the ASCII values back to string
    message = Convert_Num(decrypted_numbers)
    return message

### <span style="color:DarkRed">5.</span>  Create a Code Demo (10 points)
<span style="color:DarkRed">

Construct a demonstration of your functions above. This will be a **step-by-step** guide to using your code with a specific example that we can follow using the functions you have created above **using a mix of code and Markdown blocks**. Choose your own "Hello World" type message or a favorite quote of your choosing to code and decode yourself.
 
**This is essentially a test that demonstrates your code with a small walkthrough of how it works. You are simply calling the functions above. You are not using a main function or custom feature here.** Imagine this as a code interview scenario, and your interviewer/future boss wants a preliminary walkthrough of your work. Therefore, some text blocks are essential and helpful for your boss to follow your process. 

Note: You can change and add the type of cell block you are editing with the drop down menu above and go between Code and Markdown.

**This is a mix of code blocks and text blocks** and must discuss all three parts for full credit. You must **show and explain your method and demonstrate (by calling your function as an example)** for generating keys and how your encoding and decoding functions work. 

For full credit: 
    
* When you call FME, explain breifly why we can can't take the "mod" of our number using "%". 
* When you call the EEA, explain why this is so essential and really the key to RSA process.</span>



# Demonstrate and explain how you Generate keys with code and text blocks.

### Step 1: Choose two prime numbers.
We want to select two prime numbers that are distinct (not the same).

In [28]:
user = input() #61
p = int(user)

user_2 = input() #53
q = int(user_2)

 61
 53


### Step 2: Calculate n
Both our public key and private key will be using the value 'n' which is the product of our two prime numbers.

In [29]:
n = p * q
print(f" n is {n}")

 n is 3233


### Step 3: Calculate Public Key
We will take both of the prime numbers we declared above and pass it to our Find_Public_Key_e function.
e is calculated by first calculating phi (Euler's function) which is found by calculating (p - 1)*(q - 1). Then we call our Euclidean_Alg function which will provide us with our e.
The Euclidean Algorithm finds an e such that 1 < e < phi and the gcd of e and phi = 1.
I also had the program print the values.

In [30]:
n, e = Find_Public_Key_e(p, q)

print(f"Public Key: n = {n}, e = {e}")

Public Key: n = 3233, e = 7


### Step 4: Calculate Private Key
We will take both of the prime numbers declared above, and pass it into our Find_Private_Key_d function.
d is calculated by calculating PHI (Euler's function) which finds the number of integers that are coprime with p * q. 
Then we use the Extended Euclidean Algorithm to find d, the modular inverse of e mod phi
I also had the program print the values.

In [31]:
d = Find_Private_Key_d(e, p, q)

print(f"Private Key: n = {n}, d = {d}")

Private Key: n = 3233, d = 1783


# Demonstrate and explain how you encode a message with code and text blocks.

### Step 1: Generate RSA Public key
This was actually already done above. You will see that our Public Key (3233, 7) was generated. We will be using this to encode our message.

### Step 2: Grab user's input
We need to take in the message we want to encode.

In [33]:
message = input()

 Hello World


### Step 3: Encode user's message
We will need to use the function we created, Encode, to encode the message. 
We will pass the n and e we generated in the previous section into the function as well as the message we just received from the user.
Encode will calculate message ^ e mod n for each ASCII value
I also want to print this encoded text

In [34]:
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [1087, 3071, 1877, 1877, 3183, 2774, 1298, 3183, 1797, 1877, 2872]


# Demonstrate and explain how you decode the same message with code and text blocks.


### Step 1: Generate RSA Private key
This was also done in the first step. You will see that our Private Key (3233, 1783) was generated. We will be using that Private Key to decode our message.

### Step 2: Take the Encoded message above and pass it through our Decode function
Our decode function will take our cipher_text and use our private key to return it back to ASCII format. This is done by using cipher_text ^ d mod n. 
We will then use Convert_Num (inside of Decode) and covert it back to a readible message.

In [35]:
decoded = Decode(n, d, cipher_text)
print(f"Decoded message: {decoded}")

Decoded message: Hello World


### 6. Code Exchange Results (5 points) 
<span style="color:DarkRed"> Now that your code is working and you are able read and write messages, include 3 complete examples of exchanging code (and exchange is both reading a message and responding) from Piazza here.  Both call your code and output code exchange results here. 
We are just looking to see which 3 codes you have exchanged in Piazza.</span>

In [7]:
#This is me decoding the original message in Piazza.
#I only included decode, Convert_Num and FME in this run in order to decipher
cipher_text = [2128, 1150, 4250, 1349, 1262, 3336, 2371, 2497, 519, 1262, 1263, 1105, 3336, 1349, 1262, 2310, 1105, 3336, 4115, 762, 2405, 1263, 1105, 3336, 1262, 1349, 1150, 1105, 1262, 506, 1105, 1105, 4723, 2405, 2497, 519, 1262, 1974, 2371, 58, 1262, 519, 1105, 1349, 1262, 4839, 1150, 1105, 2497, 1262, 1974, 2371, 58, 762, 1262, 13, 4679, 1573, 1262, 4115, 2371, 2310, 1105, 1262, 2405, 3336, 1262, 4839, 2371, 762, 1560, 2405, 2497, 519, 3250]
n = 5251
d = 3403
decoded = Decode(n, d, cipher_text)
print(f"Decoded message: {decoded}")

Decoded message: What song best describes the feeling you get when your RSA code is working?


In [12]:
#This is me encoding my response back.
#To do this, I only need the functions Encode, Convert_Text, and FME
n = 5251
e = 3
message = "I think I'd choose the song Confident (rock version) by Demi Lovato because the feeling I get after getting any piece of code to work makes me feel I can take on anything."
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [443, 1262, 1349, 1150, 2405, 2497, 1560, 1262, 443, 1558, 2310, 1262, 4115, 1150, 2371, 2371, 3336, 1105, 1262, 1349, 1150, 1105, 1262, 3336, 2371, 2497, 519, 1262, 1456, 2371, 2497, 506, 2405, 2310, 1105, 2497, 1349, 1262, 988, 762, 2371, 4115, 1560, 1262, 4720, 1105, 762, 3336, 2405, 2371, 2497, 658, 1262, 1263, 1974, 1262, 4623, 1105, 3283, 2405, 1262, 3143, 2371, 4720, 4250, 1349, 2371, 1262, 1263, 1105, 4115, 4250, 58, 3336, 1105, 1262, 1349, 1150, 1105, 1262, 506, 1105, 1105, 4723, 2405, 2497, 519, 1262, 443, 1262, 519, 1105, 1349, 1262, 4250, 506, 1349, 1105, 762, 1262, 519, 1105, 1349, 1349, 2405, 2497, 519, 1262, 4250, 2497, 1974, 1262, 2911, 2405, 1105, 4115, 1105, 1262, 2371, 506, 1262, 4115, 2371, 2310, 1105, 1262, 1349, 2371, 1262, 4839, 2371, 762, 1560, 1262, 3283, 4250, 1560, 1105, 3336, 1262, 3283, 1105, 1262, 506, 1105, 1105, 4723, 1262, 443, 1262, 4115, 4250, 2497, 1262, 1349, 4250, 1560, 1105, 1262, 2371, 2497, 1262, 4250, 2497, 1974, 1349, 1150, 24

In [14]:
#This is me decoding another peers message
#I only included decode, Convert_Num and FME in this run in order to decipher
cipher_text = [146, 78, 15, 116, 210, 289, 144, 289, 210, 127, 297, 156, 210, 289, 297, 210, 293, 297, 160, 210, 107, 156, 257, 127, 210, 104, 116, 78, 241]
n = 299
d = 233
decoded = Decode(n, d, cipher_text)
print(f"Decoded message: {decoded}")

Decoded message: What did you do for July 4th?


In [37]:
#This is me encoding my response back to my peer.
#To do this, I only need the functions Encode, Convert_Text, and FME
n = 299
e = 17
message = "I spent my fourth of July working on my project for 2400."
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [255, 210, 46, 99, 95, 54, 116, 210, 57, 127, 210, 293, 297, 156, 160, 116, 78, 210, 297, 293, 210, 107, 156, 257, 127, 210, 71, 297, 160, 217, 144, 54, 129, 210, 297, 54, 210, 57, 127, 210, 99, 160, 297, 227, 95, 203, 116, 210, 293, 297, 160, 210, 163, 104, 133, 133, 115]


In [36]:
#Finally, this is how I decoded my response that used my own public/private keys
n = 3233
e = 7
message = "If you could instantly learn any skill in the world, what would it be and why?"
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [286, 576, 2774, 731, 3183, 1297, 2774, 24, 3183, 1297, 1877, 2872, 2774, 3020, 1544, 567, 1762, 1818, 1544, 1762, 1877, 731, 2774, 1877, 3071, 1818, 1797, 1544, 2774, 1818, 1544, 731, 2774, 567, 2174, 3020, 1877, 1877, 2774, 3020, 1544, 2774, 1762, 3052, 3071, 2774, 863, 3183, 1797, 1877, 2872, 1129, 2774, 863, 3052, 1818, 1762, 2774, 863, 3183, 1297, 1877, 2872, 2774, 3020, 1762, 2774, 2920, 3071, 2774, 1818, 1544, 2872, 2774, 863, 3052, 731, 2080]


### 7. Write a Narrative (25 points) 
<span style="color:DarkRed">TWO pages of meaningful narrative about your process - not including the examples -  (800 words) here in this block. Include specifc examples of debugging an challenges referencing your individual process. Responses that are vague, lacking in specific and individual details, or sound like CHatGPT will not recive full credit. Include the following:

* Describe your process. Did you plan ahead? Did you wing it? What have you learned about your own process? 
* Describe the results of exchanging keys and codes with classmates. Did it work the first time or did you need to make adjustments?<span style = "color:white">  Use the word fortitude

* What was most challenging? <span style = "color:white">  Use a cooking metaphor.

* Who helped you with your project?  Which resources were most helpful?

* What was your Best Mistake? (funniest? most frustrating? the one you learned the most from?)</span>

I am not typically the type of person to plan ahead. I tend to dive straight into tasks and figure things out as I go. I started off with the Mastery Workbook assignments which was due sooner than the RSA Project. A great thing about the most recent Mastery Workbook assignments was that it provided me with the framework to really understand how to start this project. The workbooks also had us program the Extended Euclidian Algorithm and the Fast Modular Exponentiation which I could use in this project. While doing this project, I’ve come to realize just how important it is to first sit down and organize your thoughts before jumping in and coding. In the past, I’ve spent days and weeks attempting to finish a coding assignment just because I decided to jump straight into coding without first understanding why the algorithm worked and without attempting to write pseudocode. The Mastery Workbook really had me break apart each part of the RSA algorithm and helped me understand how each function plays a part in the algorithm.  

I put off doing the exchange with my classmates till I could completely run my code with success and accuracy, so I didn’t have to make many adjustments. I made a mistake by typing in one of my keys wrong which caused a massive jumble of random characters, however, I found the typo and was able to correct it.  I took the code I created for the Encode and Decode function and put it in another file so I can switch the n, d, and e to match whatever my classmates were using. This method prevented me from making any mistakes when attempting to decode their messages. It was a lot of fun decoding the messages and answering back in an encoded message. It reminded me of when I was a child making a new language with my friends so the adults wouldn’t understand what we were talking about. 

The most challenging part of this project was the fact that I’ve never written in Python before I taken this class. All my code writing has been in C++ and/or assembly. I think this would be like trying to cook in another language. You know how cooking works and you’ve baked yourself a cake in the past, however, now it is in a whole new language you’ve never read before. You can easily understand the basic framework (you need to mix the dry ingredients, then the wet ingredients, and then combine them. Next, you would put the cake in the oven at x degrees), however, you don’t understand this particular recipe. You need to learn what “flour” means in another language. Funny enough, I would get so frustrated because I kept adding a semi-colon at the end of each of my lines out of habit and would have to go through and remove them because I kept getting compiling errors. Those were easy fixes; however, they were tedious.  
 
Most of the resources I used were within Moodle and they were things I learned from the lessons from previous weeks. For instance, I was able to write my FME function because of Algorithm 5 on page 254 within the “Discrete Mathematics and Its Applications” book by Kenneth H. Rosen. Additionally, you can find pseudocode for the Euclidean Algorithm on page 269 of the same book. One thing I would like to point out is that for the Extended Euclidean Algorithm, I went the route of creating a loop that continuously applies the Euclidean Algorithm until b is 0. But after more research, I found we could have done this using recursion (calling the function within the function). I found this information in the following site: https://www.geeksforgeeks.org/euclidean-algorithms-basic-and-extended/. They provide many resources for different algorithms, but I decided not to look there until after I figured out how I’d implement it.  

As mentioned in a previous paragraph, my biggest hurdle was learning the syntax for Python. From the lack of semi-colons, to understanding how important it is to indent correctly, it took me a lot of time to understand how the language works. Even though it was very difficult for me to learn the language, I am glad I did. I know it will be very beneficial to my career to have a few languages under my belt. I don’t think I struggled as much on learning the algorithms because I really strengthened my programming skills after taking Data Structures at CU Boulder. I learned that with a lot of those algorithms, it’s important to figure out the basics and the theory behind it before you jump straight into programming. Do the calculations by hand first and see why it works. Additionally, if I struggle with any algorithm, what I will do is find someone’s implementation of it online, print it out, and go line by line and try to determine what the output is without running it on my computer. It can really help you understand what each line is doing within the code.  

### 8. How to Break Code. (10 points)

<span style="color:DarkRed">


**This section (and section 9 and 10) of the project is only available to students exchanging codes by the Friday (midnight) before the project is due.**

**Implementing a factoring algorithm:**

Begin by coding the basic brute force factorization algorithm given in pseudocode below.

Brute Force Factoring
<pre><code>def factorize(n):
    # n is a number, return the smallest factor of n
    for i from 2 to n-1:
        if i divides n:
            return i
        return FALSE
        </code></pre>

In [20]:
def factorize(n):
    # This function attempts to find the smallest factor of n by iterating from 2 to n-1 to find the smallest factor. 
    #It checks if that value divides into n, if not, continue the loop until we reach n-1. If we do find a factor, we return that value. 
    
    for i in range(2, n): #Looping until we find a factor of n or we reach n-1. 
        if n % i == 0: #if the mod of n and the number is 0, it is a factor
            return i #We will return that value
    return False #We will return false if we reach n-1 and find no factor 

Smallest Factor of 3233 is 53


<span style="color:DarkRed"> After you have tested your RSA package yourself, and tested it with classmates by publishing both the private and public keys on Piazza, post 2 messages with just the public keys for people to break on the "Just Public Keys" thread - use both very small n ‘s (under 1000) for practice. Feel free to come up with n's that create a challenge.
        
* Explain in detail how breaking works in this section using markdown blocks. Demo a short example with a small n. What are the complete steps? What tools do you need? <span style = "color:white">  Use a sports metaphor.
* Try brute forcing a larger value of n, say 15-18 digits long. What do you notice? (If this is taking a while, just ride it out, or Ctrl+C will KeyboardInterrupt the cell block). Provide a sample large n that will take ~3-5 min to crack. 
    * Optional Challenge: Consider illustrating this with python math and time modules. matplotlib and time are good examples. Documentation is available.
* How secure is RSA? Why is it difficult to break codes in RSA? 
* Is our implementation of RSA secure? What are it's flaws? </span>

In [22]:
#Here we are just using our n from above and printing out whether we found a factor.
number = 3233
factor = factorize(number)
if factor:
    print(f"Smallest Factor of {number} is {factor}")
else:
    print(f"{number} is prime")

Smallest Factor of 3233 is 53


In [23]:
number = 1689259081189 #1299709 * 1299721 (two prime numbers)
factor = factorize(number)
if factor:
    print(f"Smallest Factor of {number} is {factor}")
else:
    print(f"{number} is prime")

Smallest Factor of 1689259081189 is 1299709


This particular way of breaking the algorithm works by first finding a factor of n. Since n is part of the public key, this is known to everyone. 
In this particular instance, n = 1689259081189, we put it through the factorize function and get 1299709, which is p. We can then find q by dividing n by the smallest factor, which will equal 1299721 (q). 
Now that we have both p and q, we can calculate the private key d as follows:

In [30]:
#First we need to find the public key to get e:
#I'm keeping this in here, but I realized that this is unecessary. We already have e from the public key.
p = 1299709
q = 1299721
n, e = Find_Public_Key_e(p, q)

print(f"Public Key: n = {n}, e = {e}")

Public Key: n = 1689259081189, e = 7


In [31]:
#Now that we have e, we can find the private key:
d = Find_Private_Key_d(e, p, q)

print(f"Private Key: n = {n}, d = {d}")

Private Key: n = 1689259081189, d = 1447934127223


Security of RSA's depend on the key length. If the length of n is very short (like in our demonstrations) it can be very easy to decode. However, RSA's used in real life applications are extremely long and very difficult to decode. It would be very inefficient if we were to use our brute force breaking on a key that is much longer. 

### 9. Code Breaking: Complete Examples (5 points)

<span style="color:DarkRed">

* Decode one example from Piazza in detail showing all the steps. Use a combination of code and markdown blocks.
* Decode and Respond to 3 "just codes with public keys" in Piazza with different sizes of n. Include complete results here - you may just show the results here.</span>


### Example 1

In [32]:
#Finding the smallest factor
number = 703
factor = factorize(number)
if factor:
    print(f"Smallest Factor of {number} is {factor}")
else:
    print(f"{number} is prime")

Smallest Factor of 703 is 19


In [60]:
#Now that we have both p and q, we can grab e from the public key and calculate the Private key
e = 5
p = 19
q = 37
d = Find_Private_Key_d(e, p, q)

print(f"Private Key: d = {d}")

Private Key: d = 389


In [47]:
#Decipher the message using decode
cipher_text = [544, 472, 233, 242, 250, 555, 591, 550, 376, 364, 210, 165, 526, 47, 555, 95, 476, 210, 242, 165, 95, 233, 526, 210, 376, 95, 233, 242, 364, 210, 242, 224, 376, 95, 364, 233, 47, 242, 210, 555, 376, 165, 472, 242, 555, 410, 242, 250, 95, 233, 233, 47, 145]
n = 703
d = 389
decoded = Decode(n, d, cipher_text)
print(f"Decoded message: {decoded}")

Decoded message: The Conquistador's treasure is buried south of Creed.


In [45]:
#My response back using encode
n = 703
e = 5
message = "I never watched Monty Python either so if it is a reference to that movie I'm unsure as well!"
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [517, 242, 591, 233, 416, 233, 95, 242, 541, 526, 165, 511, 472, 233, 47, 242, 58, 555, 591, 165, 581, 242, 302, 581, 165, 472, 555, 591, 242, 233, 364, 165, 472, 233, 95, 242, 210, 555, 242, 364, 410, 242, 364, 165, 242, 364, 210, 242, 526, 242, 95, 233, 410, 233, 95, 233, 591, 511, 233, 242, 165, 555, 242, 165, 472, 526, 165, 242, 523, 555, 416, 364, 233, 242, 517, 476, 523, 242, 376, 591, 210, 376, 95, 233, 242, 526, 210, 242, 541, 233, 90, 90, 86]


### Example 2

In [48]:
#Finding the smallest factor
number = 629
factor = factorize(number)
if factor:
    print(f"Smallest Factor of {number} is {factor}")
else:
    print(f"{number} is prime")

Smallest Factor of 629 is 17


In [61]:
#Now that we have both p and q, we can grab e from the public key and calculate the Private key
e = 5
p = 17
q = 37
d = Find_Private_Key_d(e, p, q)

print(f"Private Key: d = {d}")

Private Key: d = 461


In [54]:
#Decipher the message
cipher_text = [280, 156, 165, 287, 492, 79, 156, 165, 549, 437, 506, 427, 549, 506, 427, 165, 287, 492, 427, 79, 376, 506, 549, 437, 427, 518, 595, 427, 428, 492, 156, 506, 518, 332, 71, 427, 282, 427, 296, 156, 79, 492, 506, 427, 296, 518, 506, 492, 482, 287, 427, 478, 100, 534, 305, 492, 506, 165, 492, 428]
n = 629
d = 461
decoded = Decode(n, d, cipher_text)
print(f"Decoded message: {decoded}")

Decoded message: Mathematics is the music of reason. - James Joseph Sylvester


In [55]:
#My response back
n = 629
e = 5
message = "Pretty cool quote. I am quite a fan of music so it's interesting to see the comparison."
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [598, 428, 492, 165, 165, 100, 427, 437, 518, 518, 534, 427, 180, 376, 518, 165, 492, 71, 427, 184, 427, 156, 79, 427, 180, 376, 549, 165, 492, 427, 156, 427, 595, 156, 332, 427, 518, 595, 427, 79, 376, 506, 549, 437, 427, 506, 518, 427, 549, 165, 439, 506, 427, 549, 332, 165, 492, 428, 492, 506, 165, 549, 332, 273, 427, 165, 518, 427, 506, 492, 492, 427, 165, 287, 492, 427, 437, 518, 79, 482, 156, 428, 549, 506, 518, 332, 71]


### Example 3

In [57]:
#Finding the smallest factor
number = 559
factor = factorize(number)
if factor:
    print(f"Smallest Factor of {number} is {factor}")
else:
    print(f"{number} is prime")

Smallest Factor of 559 is 13


In [62]:
#Now that we have both p and q, we can grab e from the public key and calculate the Private key
e = 5
p = 13
q = 43
d = Find_Private_Key_d(e, p, q)

print(f"Private Key: d = {d}")

Private Key: d = 101


In [64]:
#Decipher the message
cipher_text = [501, 24, 52, 457, 234, 145, 27, 511, 457, 314, 24, 12, 457, 150, 145, 417, 426, 511, 367, 457, 52, 314, 12, 417, 426, 457, 127, 24, 52, 457, 237, 52, 417, 12, 457, 12, 134, 127, 417, 314, 298, 544]
n = 559
d = 101
decoded = Decode(n, d, cipher_text)
print(f"Decoded message: {decoded}")

Decoded message: You have not failed until you quit trying.


In [66]:
#My response back
n = 559
e = 5
message = "True! Especially with coding, you have to keep trying until you get it."
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [54, 134, 52, 511, 362, 457, 218, 20, 476, 511, 203, 417, 145, 426, 426, 127, 457, 448, 417, 12, 234, 457, 203, 24, 367, 417, 314, 298, 44, 457, 127, 24, 52, 457, 234, 145, 27, 511, 457, 12, 24, 457, 477, 511, 511, 476, 457, 12, 134, 127, 417, 314, 298, 457, 52, 314, 12, 417, 426, 457, 127, 24, 52, 457, 298, 511, 12, 457, 417, 12, 544]


### Generating two public keys along with messages

In [67]:
#Public key #1
p = 2
q = 9
n, e = Find_Public_Key_e(p, q)

print(f"Public Key: n = {n}, e = {e}")

Public Key: n = 18, e = 3


In [68]:
#Encode message
n = 18
e = 3
message = "The only limit to our realization of tomorrow is our doubts of today. — Franklin D. Roosevelt"
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [0, 8, 17, 8, 9, 8, 0, 1, 8, 0, 9, 1, 9, 8, 8, 8, 9, 8, 9, 9, 0, 8, 0, 17, 1, 0, 9, 8, 1, 8, 9, 9, 8, 8, 9, 0, 8, 8, 9, 1, 9, 0, 0, 9, 17, 8, 9, 1, 8, 9, 9, 0, 8, 10, 9, 9, 8, 8, 1, 8, 9, 0, 8, 8, 9, 10, 1, 1, 10, 8, 10, 8, 10, 0, 1, 8, 17, 0, 9, 8, 8, 8, 10, 8, 10, 9, 9, 1, 17, 10, 17, 0, 8]


In [69]:
#Public Key #2
p = 27
q = 17
n, e = Find_Public_Key_e(p, q)

print(f"Public Key: n = {n}, e = {e}")

Public Key: n = 459, e = 3


In [70]:
#Encode message
n = 459
e = 3
message = "Life is what happens when you're busy making other plans. — John Lennon"
cipher_text = Encode(n, e, message)
print(f"Encoded Message: {cipher_text}")

Encoded Message: [172, 27, 0, 305, 179, 27, 208, 179, 170, 314, 181, 296, 179, 314, 181, 388, 388, 305, 359, 208, 179, 170, 314, 305, 359, 179, 280, 270, 162, 108, 351, 305, 179, 242, 162, 208, 280, 179, 190, 181, 431, 27, 359, 307, 179, 270, 296, 314, 305, 351, 179, 388, 216, 181, 359, 208, 28, 179, 307, 179, 386, 270, 314, 359, 179, 172, 305, 359, 359, 270, 359]



### 10. Custom CODE Feature  (10 points) ##
<span style="color:DarkRed">- The custom feature should be a "Stand Alone" feature at the end of the notebook. 
- If you need  to use code from above, please copy the entire package and functions again as part of your feature so it can be tested independently. 
- You may remove the give code comments in this section.
- Be sure to provide a demonstration showing that it works. See RSA Guide for more information and suggestions. 
- Include a narrative describing your project.
- All custom features must include a narrative explaining the features and all code must be commented.</span>

**Do not forget the self-grading element.**


**Please use the grading guide below to self grade your Custom Feature.**

GRADING FOR CUSTOM FEATURE:

10 pts  Wow.  It is amazing.
        Shows initiative and originality. 
        You did something extra special and pushed yourself.
        You went beyond all expectations 
        You broke the rules in a creative way. 
        Your coding and commenting is exceptional.
        Excellent 

7-9 points - It is GOOD! 
        You were a “self starter.” 
         You did everything  requested. 
        All expectations met. 
       	You did a very good job. 
        Good use of commenting.
        Shows mastery of skills.

5-6 points  - Okay.
        Minimum requirements.
        Commenting weak.

0-4 points - It is not finished. 
            Does meet objectives. 
            Did not follow directions. 
            Chose not to do this part (which is a totally fine choice)


## Self grading of custom feature ##

Here in this block, please rate what score out of 8 your custom feature merits given the criteria above:

* For example, if you are brand new to programming, using a main function could be an amazing feature - tell me how this helped you move to a new level.
* In what ways did you push yourself? Did you try something you've never done before?
* Alternatively, if you just did not have the time to do the custom feature, this is actually just fine and a reasonable choice. No judgment.

0 - I decided not to do this part. I didn't have time with juggling a full time job and a family. But, it might be cool to pick this up another time to explore different options!